File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit_objc/Protocols/FLTTransactionCacheProtocol.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "../include/in_app_purchase_storekit_objc/FLTTransactionCacheProtocol.h"

@interface DefaultTransactionCache ()
/// The wrapped FIATransactionCache
@property(nonatomic, strong) FIATransactionCache *cache;
@end

@implementation DefaultTransactionCache

- (void)addObjects:(nonnull NSArray *)objects forKey:(TransactionCacheKey)key {
  [self.cache addObjects:objects forKey:key];
}

- (void)clear {
  [self.cache clear];
}

- (nonnull NSArray *)getObjectsForKey:(TransactionCacheKey)key {
  return [self.cache getObjectsForKey:key];
}

- (nonnull instancetype)initWithCache:(nonnull FIATransactionCache *)cache {
  self = [super init];
  if (self) {
    _cache = cache;
  }
  return self;
}
@end

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit/Protocols/FLTTransactionCacheProtocol.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

/// A protocol that defines a cache of all transactions, both completed and in progress.
@objc public protocol FLTTransactionCacheProtocol: NSObjectProtocol {
  /// Adds objects to the transaction cache.
  ///
  /// If the cache already contains an array of objects on the specified key, the supplied
  /// array will be appended to the existing array.
  func add(_ objects: [Any], for key: TransactionCacheKey)

  /// Gets the array of objects stored at the given key.
  ///
  /// If there are no objects associated with the given key an empty array is returned.
  func getObjectsFor(_ key: TransactionCacheKey) -> [Any]

  /// Removes all objects from the transaction cache.
  func clear()
}

/// The default transaction cache that wraps FIATransactionCache
public class DefaultTransactionCache: NSObject, FLTTransactionCacheProtocol {
  /// The wrapped FIATransactionCache
  private let cache: FIATransactionCache

  /// Initialize this wrapper with an FIATransactionCache
  public init(cache: FIATransactionCache) {
    self.cache = cache
  }

  public func add(_ objects: [Any], for key: TransactionCacheKey) {
    cache.add(objects, for: key)
  }

  public func getObjectsFor(_ key: TransactionCacheKey) -> [Any] {
    return cache.getObjectsFor(key)
  }

  public func clear() {
    cache.clear()
  }
}
